File: JuliaCode/TestImageProcessing.m
% Test Image Processing
% Generates files for verification of algorithms in
% `JuliaImageProcessing.jl`.
% References:
%   1.  
% Remarks:
%   1.  B
% TODO:
% 	1.  C
% Release Notes Royi Avital RoyiAvital@yahoo.com
% - 1.0.000     07/09/2024
%   *   First release.


%% General Parameters

subStreamNumberDefault = 79;

run('InitScript.m');

figureIdx           = 0;
figureCounterSpec   = '%04d';

generateFigures = ON;

%% Simulation Constants

% Match Julia
PAD_MODE_CIRCULAR   = 0;
PAD_MODE_CONSTANT   = 1;
PAD_MODE_REFLECT    = 2;
PAD_MODE_REPLICATE  = 3;
PAD_MODE_SYMMETRIC  = 4;

% Match Julia
CONV_MODE_FULL  = 0;
CONV_MODE_SAME  = 1;
CONV_MODE_VALID = 2;


%% Simulation Parameters

numRows = 26;
numCols = 25;

repFactor = 15;

vPadMode    = [PAD_MODE_CIRCULAR; PAD_MODE_CONSTANT; PAD_MODE_REPLICATE; PAD_MODE_SYMMETRIC];
vConvMode   = [CONV_MODE_FULL; CONV_MODE_SAME; CONV_MODE_VALID];

vPadMode = repmat(vPadMode, repFactor, 1);
vConvMode = repmat(vConvMode, repFactor, 1);


%% Generate / Load Data

mI = rand(numRows, numCols);


%% Pad Array

cPad = {length(vPadMode), 3}; %<! Output, vPadSize, PadMode

for ii = 1:length(vPadMode)
    vPadSize = randi([1, 9], 1, 2);
    switch(vPadMode(ii))
        case(PAD_MODE_CIRCULAR)
            padMode = 'circular';
        case(PAD_MODE_CONSTANT)
            padMode = 'constant';
        case(PAD_MODE_REPLICATE)
            padMode = 'replicate';
        case(PAD_MODE_SYMMETRIC)
            padMode = 'symmetric';
    end
    if (vPadMode(ii) == PAD_MODE_CONSTANT)
        mIPad = padarray(mI, vPadSize, 0, 'both');
    else
        mIPad = padarray(mI, vPadSize, padMode, 'both');
    end
    cPad{ii, 1} = mIPad;
    cPad{ii, 2} = vPadSize;
    cPad{ii, 3} = padMode;
end


%% Convolution

cConv = {length(vConvMode), 4}; %<! Output, vKerSize, mK, convMode

for ii = 1:length(vConvMode)
    vKerSize = randi([1, 9], 1, 2);
    mK = rand(vKerSize); %<! Must be row vector
    switch(vConvMode(ii))
        case(CONV_MODE_FULL)
            convMode = 'full';
        case(CONV_MODE_SAME)
            convMode = 'same';
        case(CONV_MODE_VALID)
            convMode = 'valid';
    end
    mIConv = conv2(mI, mK, convMode);
    cConv{ii, 1} = mIConv;
    cConv{ii, 2} = vKerSize;
    cConv{ii, 3} = mK;
    cConv{ii, 4} = convMode;
end


%% Save Data

save('TestImageProcessing', 'mI', 'cPad', 'cConv');



%% Auxiliary Functions



%% Restore Defaults

% set(0, 'DefaultFigureWindowStyle', 'normal');
% set(0, 'DefaultAxesLooseInset', defaultLoosInset);

File: JuliaCode/TestImageProcessing.jl
# Test Code for Image Processing
# Several test for the Julia Code.
# References:
#   1.  
# Remarks:
#   1.  Use in Julia as following:
#       -   Move to folder using `cd(raw"<PathToFolder>");`.
#       -   Activate the environment using `] activate .`.
#       -   Instantiate the environment using `] instantiate`.
#   2.  fd
# TODO:
# 	1.  C
# Release Notes Royi Avital RoyiAvital@yahoo.com
# - 1.0.000     07/09/2024  Royi Avital
#   *   First release.

## Packages

# Internal
using Printf;
# External
using BenchmarkTools;
using MAT;
using UnicodePlots;

## Constants & Configuration

## External
include("JuliaInit.jl");
include("JuliaImageProcessing.jl");

## General Parameters

figureIdx = 0;

exportFigures = false;

## Functions


## Parameters

# Data
matFileName = "TestImageProcessing.mat";

dPadMode = Dict("circular" => PAD_MODE_CIRCULAR, "constant" => PAD_MODE_CONSTANT, "replicate" => PAD_MODE_REPLICATE, "symmetric" => PAD_MODE_SYMMETRIC);
dConvMode = Dict("full" => CONV_MODE_FULL, "same" => CONV_MODE_SAME, "valid" => CONV_MODE_VALID);


## Load / Generate Data

dMatData = matread(matFileName);
mI = dMatData["mI"];
numRows, numCols = size(mI);


## Analysis

# Pad Array
cPad = dMatData["cPad"]; #<! Cell array is mapped to a vector of `Any` in Julia
numTests = size(cPad, 1);

println("Pad Array Tests")
for ii ∈ 1:numTests
    tuPadSize   = tuple(Int.(cPad[ii, 2])...);
    padMode     = dPadMode[cPad[ii, 3]];
    
    mO = PadArray(mI, tuPadSize, padMode);
    
    maxErr = maximum(abs.(mO - cPad[ii, 1]));
    @printf("Iteration: %04d, Pad Size: (%d, %d), Pad Mode: %s, Maximum Error: %0.3f\n", ii, tuPadSize[1], tuPadSize[2], cPad[ii, 3], maxErr);
end

println("")

# Convolution 2D
cConv = dMatData["cConv"]; #<! Cell array is mapped to a vector of `Any` in Julia
numTests = size(cConv, 1);

println("Convolution 2D Tests")
for ii ∈ 1:numTests
    tuKerSize   = tuple(Int.(cConv[ii, 2])...);
    # Handle the case the MATLAB matrix is 1x1 which is interpreted in Julia as a scalar
    mK          = Matrix{Float64}(undef, tuKerSize);
    mK[:]       = collect(cConv[ii, 3])[:];
    convMode    = dConvMode[cConv[ii, 4]];

    mO = Conv2D(mI, mK; convMode = convMode);
    
    maxErr = maximum(abs.(mO - cConv[ii, 1]));
    @printf("Iteration: %04d, Kernel Size: (%d, %d), Conv Mode: %s, Maximum Error: %0.25f\n", ii, tuKerSize[1], tuKerSize[2], cConv[ii, 4], maxErr);
end



## Display Results

